### PR TITLE
Autoselect appropriate CI test and Scoring method for causal discovery

### LINKED ISSUE #1873: Automatically choose the CI test and scoring method depending on variable types
The current default scoring and CI test set in PC, Hill-Climb, or GES algorithm is for discrete variable CI test or scoring methods. As we now also have support for continuous and mixed variables, the default should automatically choose the appropriate CI test or scoring method depending on the variable types in the passed dataset. 

Supported data types for CI tests:
1. pearsonr and gcm: All continuous variables
2. pillai_trace: Any combination
3. Remaining: all discrete variables.

Supported data types for scoring methods:
1. K2, BDeu, BDs, BIC, AIC: All discrete variables
2. LoglikelihoodGauss, AICGauss, BICGauss: All continuous variables
3. LoglikelihoodCondGauss, AICCondGauss, BICCondGauss: Any combination

### PR DESCRIPTION
**Pull request by arocapedro**
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #1873

### List of changes to the codebase in this pull request
- Rename method `get_ci_test` to `get_callable_ci_test`, since it obtains a callable from None or string
- CI test is not automatically determined depending on the type of data (discrete or continuous)
- All the estimate methods do not have a scoring method by default, and it is determined on runtime depending on the data. For this 'get_scoring_method' is also modified.
- For passing categorical data that in numerical form, user must specify the dtype as "category" when building the dataframe. An heuristic method is used to determine if the user is making this mistake, and a warning will be raised.
- Tests are updated to address the above issue

### CODE REVIEW

File: pgmpy/estimators/CITests.py [hunk 1]
```diff
@@ -2,32 +2,88 @@
...
 from sklearn.cross_decomposition import CCA
-
+from collections.abc import Callable
+from typing import Union, Optional
 from pgmpy.global_vars import logger
 from pgmpy.independencies import IndependenceAssertion
+from pgmpy.utils.utils import check_variable_type
+
+
+def get_callable_ci_test(
```
Comment (arocapedro) [inline pgmpy/estimators/CITests.py:14]: rename this method to reflect more what it is doing: obtaining a callable from string or null

File: pgmpy/estimators/CITests.py [hunk 2]
```diff
@@ -764,3 +820,43 @@ def gcm(X, Y, Z, data, boolean=True, **kwargs):
             return False
     else:
         return t_stat, p_value
+
+
+# def select_CI_test(data: pd.DataFrame) -> tuple[str, list[str]]:
```
Comment (arocapedro) [inline pgmpy/estimators/CITests.py]: needs to be deleted

File: pgmpy/utils/utils.py [hunk 3]
```diff
@@ -415,3 +415,31 @@ def preprocess_data(df):
         f" Datatype (N=numerical, C=Categorical Unordered, O=Categorical Ordered) inferred from data: \n {dtypes}"
     )
     return (df, dtypes)
+
+
+def check_variable_type(data: pd.DataFrame) -> str:
```
Comment (ankurankan) [inline pgmpy/utils/utils.py]: The `preprocess_data` method is quite similar to this. Can we somehow use that?

Comment (arocapedro) [reply pgmpy/utils/utils.py]: I have updated [`check_variable_type`](https://github.com/arocapedro/pgmpy/blob/1030d6938516527066e5bcde046ee6a6253490a3/pgmpy/utils/utils.py#L419) to use `preprocess_data`, let me know if that's better

**Comment by ankurankan** [comment]
Thanks for the PR @arocapedro. I have made a couple of very minor changes: 1. Rename `check_variable_type` -> `get_dataset_type`. 2. Added `get_dataset_type` to the `__init__` file so that we can do `from pgmpy.utils import get_dataset_type`.
